src/heartbeat.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "heartbeat.h"
+
+#include <time.h>
+#include <uv.h>
+
+#include "icmp.h"
+#include "ip.h"
+#include "log.h"
+
+static void
+on_request_timer(uv_timer_t* timer);
+
+static void
+on_reply_timer(uv_timer_t* timer);
+
+static void
+send_heartbeat(struct enftun_heartbeat* heartbeat);
+
+static void
+on_write(struct enftun_crb* crb)
+{
+    if (crb->status)
+        enftun_log_error("PING: failed to send heartbeat reply: %d\n",
+                         crb->status);
+}
+
+static void
+send_heartbeat(struct enftun_heartbeat* heartbeat)
+{
+    heartbeat->hb_scheduled = false;
+    heartbeat->hb_inflight  = true;
+
+    enftun_log_debug("Ping.....");
+
+    enftun_packet_reset(&heartbeat->reply_pkt);
+    enftun_icmp6_echo_request(&heartbeat->reply_pkt, heartbeat->source_addr,
+                              heartbeat->dest_addr);
+    enftun_crb_write(&heartbeat->reply_crb, heartbeat->chan);
+
+    uv_timer_start(&heartbeat->reply_timer, on_reply_timer,
+                   heartbeat->heartbeat_timeout, 0);
+}
+
+static void
+schedule_heartbeat(struct enftun_heartbeat* heartbeat)
+{
+    heartbeat->hb_scheduled = true;
+
+    if (!heartbeat->hb_inflight)
+        send_heartbeat(heartbeat);
+}
+
+static void
+on_request_timer(uv_timer_t* timer)
+{
+    struct enftun_heartbeat* heartbeat = timer->data;
+    schedule_heartbeat(heartbeat);
+}
+
+static void
+on_reply_timer(uv_timer_t* timer)
+{
+    struct enftun_heartbeat* heartbeat = timer->data;
+    heartbeat->on_timeout(heartbeat->data);
+    return;
+}
+
+int
+enftun_heartbeat_start(struct enftun_heartbeat* heartbeat)
+{
+    int rc = uv_timer_start(&heartbeat->request_timer, on_request_timer,
+                            heartbeat->heartbeat_period, 0);
+
+    return rc;
+}
+
+int
+enftun_heartbeat_restart(struct enftun_heartbeat* heartbeat)
+{
+    if (!heartbeat->hb_inflight)
+    {
+        enftun_heartbeat_stop(heartbeat);
+        uv_timer_start(&heartbeat->request_timer, on_request_timer,
+                       heartbeat->heartbeat_period, 0);
+    }
+
+    return 0;
+}
+
+int
+enftun_heartbeat_now(struct enftun_heartbeat* heartbeat)
+{
+    schedule_heartbeat(heartbeat);
+    return 0;
+}
+
+int
+enftun_heartbeat_stop(struct enftun_heartbeat* heartbeat)
+{
+    uv_timer_stop(&heartbeat->request_timer);
+    uv_timer_stop(&heartbeat->reply_timer);
+
+    if (heartbeat->hb_inflight)
```

Comment:
`inflight` is tracking two things now, has the async write completed (on_write been called) and has the reply been received or timed out (handle_packet or on_reply_timer been called).

We should only call `crb_cancel` if the former hasn't completed, not if just the latter is still outstanding.

So maybe change this to three booleans, `scheduled`, `sending`, and `inflight`.

The old `inflight` is renamed to `sending` and gets cleared when `on_write` callback is called.
The new `inflight` gets cleared when the response/timeout is received.

In `ndp.c`, also rename `inflight` to `sending`, to match the new naming scheme here.